
rm(list = ls())

library(GenomicRanges)
library(data.table)
library(parallel)
library(Segvis)

load( file = "data/RData/unified_lists_wProbs.RData")
load( file = "data/RData/summits.RData")
 
source("R/meta_functions.R")
 
peaks <- unified_lists$peaks
overlaps <- unified_lists$overlaps
probs <- unified_lists$probs
 
window_ext <- 2000
fragLen <- 200
bandwidth <- 151
mc <- detectCores()
 
histone_dir <- "inst/histone_old"
 
## only use data generated by Bernstein lab
files <- list.files(histone_dir)
files <- files[grep("tagAlign",files,invert = TRUE)]
files <- files[grep("bai",files,invert = TRUE)]
files <- files[grep("bam",files)]


sets <- c("EBNA2","EBNA3A","EBNA3B","EBNA3C","RBPJ")

create_profile_sets <- function(set,summits,window_ext)
{
  col <- summits[[paste0(set,".summit")]]
  idx <- which(!is.na(col))
  out <- GRanges(seqnames = as.character(summits[idx,(seqnames)]),
                 ranges = IRanges(start = col[idx] - window_ext,
                   end = col[idx] +  window_ext),strand = "*")
  return(out)
}
 
regions <- lapply(sets,create_profile_sets,summits,window_ext)
names(regions) <- sets
  
base <- buildSegvis(name = "",file = "", maxBandwidth = 201,chr = "human",fragLen = fragLen)
 
segvis <- lapply(sets,
                 function(x,base){
                   name(base) <- x
                   return(base)},base)
names(segvis) <- sets

segvis <- mapply(function(seg,reg){
  regions(seg) <- reg
  return(seg)},segvis,regions,SIMPLIFY = FALSE)

segvis_histone <- lapply(segvis,function(x,files,dir){
  out <- lapply(file.path(dir,files),function(ff,x){
    file(x) <- ff
    return(x)},x)
  return(out)},files,histone_dir)

## load reads
segvis_histone <- lapply(segvis_histone,
  function(x,mc){
    out <- lapply(x,loadReads,mc)
    return(out)},mc)

segvis_histone <- lapply(segvis_histone,
  function(x,mc){
    out <- lapply(x,matchReads,mc)
    return(out)},mc)

segvis_histone <- lapply(segvis_histone,
  function(x,mc){
    out <- lapply(x,getCoverage,mc)
    return(out)},mc)

depths <- mclapply(segvis_histone[[1]],
  countReads,mc.cores = mc)

segvis_block_histone <- lapply(segvis_histone,
  function(x,bw,mc){
    out <- lapply(x,Segvis_block,bandwidth,mc)
    return(out)},bw = bandwidth,mc = mc)

segvis_block_histone <- lapply(segvis_block_histone,
  function(x,depths){
    out <- mapply(function(seg,dep){
      normConst(seg) <- dep
      seg <- normalize(seg)
      return(seg)},x,depths,SIMPLIFY = FALSE)
    return(out)},depths)
    
## tab <- meta[,.(File.accession, Experiment.target)]
## setkey(tab,File.accession)
## tab[,repl := 1L]
## tab[,Experiment.target := gsub("-human","",Experiment.target)]

## tab <- tab[gsub(".bam","",files)]
## aux <- tab[,length(repl),by = Experiment.target]
## setkey(tab,Experiment.target)
## setkey(aux,Experiment.target)

## for(h in aux[,(Experiment.target)]){
##   reps <- aux[h,(V1)]
##   tab[h , repl := 1:reps]
## }

## tab[,name := paste0(Experiment.target,"_rep",repl)]
## nms <- tab[,(name)]

nms <- gsub("wgEncodeBroadHistoneGm12878","",files)
nms <- gsub("StdAlnRep","-",nms)
nms <- gsub(".bam","",nms)


segvis_block_list <- lapply(segvis_block_histone,
  function(x,nms){
    out <- Segvis_block_list(x)
    names(out) <- nms
    return(out)},nms)

out_dir <- "data/RData/profile_matrices_old"

for(set in sets){
  for(nm in nms){
    profile <- cover_table(segvis_block_list[[set]][[nm]])
    save(profile, file = file.path(out_dir,paste0(set,"_",nm,"_profile.RData")))
  }
}
  
## extract_fileAcc <- function(segvis)
## {
##   file <- file(segvis)
##   file <- strsplit(file,"/")[[1]][3]
##   file <- gsub(".bam","",file)
##   return(file)
## }



## segvis_histone <- lapply(segvis_histone,
##  function(x,nms){
##    names(x) <- nms
##    return(x)},nms)


## k <- 1
## aux <- list()
## for(set in sets){
##   for(nm in nms){
##     ff <- extract_fileAcc(segvis_histone[[set]][[nm]])
##     z <- copy(tab[ff])
##     z[,SS := set]
##     aux[[k]] <- z
##     k <- k+1
##     }
## }

## aux <- do.call(rbind,aux)



